fix(ssa): Validate signed mul overflow

vezenovm opened the pull request:
# Description

## Problem\*

Resolves #8088 

Builds upon https://github.com/noir-lang/noir/pull/8765.

## Summary\*

This adds validation similar to https://github.com/noir-lang/noir/pull/8706, except for signed multiplication. The semantics for which are seen here. We could be more strict with these checks by checking this code gen as well:
```rust
self.check_signed_overflow(result, lhs, rhs, operator, bit_size, location);
self.insert_safe_cast(result, result_type, location)
```
However, for now I have chosen to simply check up to the truncate point to simplify our "state machine". We also mainly want to guarantee that we have truncated the result of an operation appropriately so that someone cannot code gen an overflow.

## Additional Context

## Documentation\*

Check one:
- [ ] No documentation needed.
- [X] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

asterite commented on compiler/noirc_evaluator/src/ssa/parser/into_ssa.rs:
```diff
@@ -558,10 +557,6 @@ impl Translator {
         // before each print.
         ssa.normalize_ids();
 
-        for function in ssa.functions.values() {
-            validate_function(function);
-        }
```
Why is this gone? (maybe it's already done somewhere else and this was redundant?)

asterite reviewed (approved):
Looks good? Would a test for `mul` followed by `truncate` but not by any cast be worth having?

vezenovm commented:
> Would a test for `mul` followed by `truncate` but not by any cast be worth having?

Yes, I will add one.

vezenovm commented:
Added a test for a mul followed by a truncate with no cast. I also added a test that a lone truncate compiles:
```
        acir(inline) pure fn main f0 {
          b0(v0: i16):
            v1 = truncate v0 to 8 bits, max_bit_size: 8
            return v1
        }
```
